emission/distribution.gno

```diff
@@ -0,0 +1,154 @@
+// ChangeDistributionPct02 changes the distribution percentage for the given two targets.
+func ChangeDistributionPct02(
+	target01 emissionTarget, pct01 uint64,
+	target02 emissionTarget, pct02 uint64,
+) {
+	changeDistributionPct(target01, pct01)
+	changeDistributionPct(target02, pct02)
+
+	checkSumDistributionPct()
+}
+
+// ChangeDistributionPct03 changes the distribution percentage for the given three targets.
+func ChangeDistributionPct03(
+	target01 emissionTarget, pct01 uint64,
+	target02 emissionTarget, pct02 uint64,
+	target03 emissionTarget, pct03 uint64,
+) {
+	changeDistributionPct(target01, pct01)
+	changeDistributionPct(target02, pct02)
+	changeDistributionPct(target03, pct03)
+
+	checkSumDistributionPct()
+}
+
+// ChangeDistributionPct04 changes the distribution percentage for the given four targets.
+func ChangeDistributionPct04(
+	target01 emissionTarget, pct01 uint64,
+	target02 emissionTarget, pct02 uint64,
+	target03 emissionTarget, pct03 uint64,
+	target04 emissionTarget, pct04 uint64,
+) {
+	changeDistributionPct(target01, pct01)
+	changeDistributionPct(target02, pct02)
+	changeDistributionPct(target03, pct03)
+	changeDistributionPct(target04, pct04)
+
+	checkSumDistributionPct()
+}
+
+// calculateAmount calculates the amount based on the given percentage in basis points.
+func calculateAmount(amount, bptPct uint64) uint64 {
+	return amount * bptPct / 10000
+}
+
+// transferToTarget transfers the specified amount to the given addresses.
+func transferToTarget(target emissionTarget, amount uint64) {
+	switch target {
+	case LIQUIDITY_STAKING:
+		// transfer to staker contract
+		gns.Transfer(a2u(consts.STAKER_ADDR), amount)
+	case DEVOPS:
+		// transfer to devops
+		gns.Transfer(a2u(consts.DEV_OPS), amount)
+	case COMMUNITY_POOL:
+		// TBD, transfer to community pool
+		gns.Transfer(a2u(consts.ZERO_ADDRESS), amount)
+	case XGNS:
+		// TBD, transfer to xGNS
+		gns.Transfer(a2u(consts.ZERO_ADDRESS), amount)
+	default:
+		panic("invalid target")
+	}
+}
+
+// changeDistributionPct changes the distribution percentage for the given target.
+func changeDistributionPct(target emissionTarget, pct uint64) {
+	// only admin or governance can change
+	caller := std.PrevRealm().Addr()
+	if caller != consts.GNOSWAP_ADMIN && caller != consts.GOV_ADDR {
+		panic("only admin or governance can change distribution percentages")
+	}
+
+	// cannot add new target
+	if target != LIQUIDITY_STAKING && target != DEVOPS && target != COMMUNITY_POOL && target != XGNS {
+		panic("invalid target")
+	}
+
+	// Maximum pct for a single target is 10000 basis points (100%)
+	if pct > 10000 {
+		panic("percentage too high")
+	}
+
+	distributionPct[target] = pct
+}
+
+// checkSumDistributionPct ensures the sum of all distribution percentages is 100%
+func checkSumDistributionPct() {
+	sum := uint64(0)
+	for _, pct := range distributionPct {
+		sum += pct
+	}
+
+	if sum != 10000 {
+		panic(ufmt.Sprintf("sum of all pct should be 100%% (10000 bps), got %d\n", sum))
```

Comment:
Panic automatically includes newlines.

Reply:
AFAIK, gno panic doesn't include new lines.

@notJoon Could you confirm this?